### DISCUSSION TITLE
generating .pb.h files with cmake?

### DISCUSSION THREAD

**Discussion by ctriley** [google/or-tools#3265] [Build and configuration issues] [answered]
TAG: stable
Language: C++

**SICP**

**ARM Mac 12.3 (21E230)?**

**What did you do?**
I'm trying to use cmake's `FetchContent` to include or-tools in a C++ library. My project structure is as follows:
```
├── CMakeLists.txt
├── src
│   ├── CMakeLists.txt # creates the dynamic library
│   ├── [various .cc/.h files]
│   ├── staticlib1
│   │   ├── CMakeLists.txt # creates a static library
│   │   └── [various .cc/.h files]
│   └── staticlib2
│       ├── CMakeLists.txt # creates another static library
│       └── [various .cc/.h files]
└── main.cc
```

In the top level `CMakeLists.txt` I have:

```cmake
    set(FETCHCONTENT_QUIET OFF)
    include(FetchContent)
    FetchContent_Declare(
            ortools
            GIT_REPOSITORY https://github.com/google/or-tools.git
            GIT_TAG        stable
    )
    set(BUILD_SHARED_LIBS ON)
    set(BUILD_DEPS ON)
    set(BUILD_SAMPLES OFF)
    set(BUILD_EXAMPLES OFF)
    FetchContent_MakeAvailable(ortools)
    add_definitions(-DORTOOLS)
```

Because of this structure, I need to include the or-tools header files in the static libraries, and then link the or-tools library against the dynamic library. Trying to link or-tools against both the static and dynamic libraries produces the following error:

ERROR: Something is wrong with flag 'flagfile' in file 'Users/username/project-root/cmake-build-debug/ortools-download/ortools_project-prefix/src/ortools_project-build/_deps/absl-src/absl/flags/parse.cc'. One possibility: file 'Users/username/project-root/cmake-build-debug/ortools-download/ortools_project-prefix/src/ortools_project-build/_deps/absl-src/absl/flags/parse.cc' is being linked both statically and dynamically into this executable. e.g. some files listed as srcs to a test and also listed as srcs of some shared lib deps of the same test.

Therefore, in the `src/staticlib1/CMakeLists.txt` I added all the header files for the dependencies:
```cmake
include_directories(${ortools_SOURCE_DIR})
include_directories(${absl_SOURCE_DIR})
...
```

and instead of linking each static library to `ortools::ortools` I just link it at the top level dynamic library:
```cmake
add_library(${dynamiclib} SHARED ${SOURCE_FILES_SRC})
target_link_libraries(${dynamiclib} ${staticlib1} ${staticlib2} ortools::ortools)
```

This seems to be working so far, but I get stuck because I don't know how to generate the `.pb.h` files (e.x. `ortools/linear_solver/linear_solver.pb.h`) and end up with an error generating the static libraries:

/Users/username/project/cmake-build-debug/_deps/ortools-src/ortools/linear_solver/linear_solver.h:159:10: fatal error: 'ortools/linear_solver/linear_solver.pb.h' file not found

Is there a way to generate these files with cmake so that I can use `include_directories` to include them in as headers in the static libraries?

**Comment by Mizux** [comment]
These file are generated by CMake using protoc
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/cmake/cpp.cmake#L146-L199
and they should be located in
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/cmake/cpp.cmake#L183
so I suspect an issue in the `target_include_directories()` of `ortools::ortools` target

dev Note:
ortools::ortools include directories
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/cmake/cpp.cmake#L92-L94

link between `ortools_proto` and `ortools` targets.
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/cmake/cpp.cmake#L216-L220

**Comment by Mizux** [comment]
upon reflexion, I think the issue is the `ortools_proto` aka `ortools::proto` is not build before you staticlib1 thus header are still not available.

you may try to add `add_dependencies(${staticlib1} ortools::proto)` to force the build of ortools::proto before staticlib1 aka execute the generation of the pb.h earlier.

half off topic: I have difficulty to exactly understand why `target_link_libraries(${staticlib1} PRIVATE ortools::ortools)` would be a problem.
which CMake generator are you using on your Mac ? (note if you don't set one explicitly you can use `cmake -G` to display the default (`*`) one)

**Comment by ctriley** [comment]
Yeah I'm not really sure what I'm doing incorrectly. 

I changed my code so that I'm only calling `target_link_libraries(lib PRIVATE ortools::ortools)` once. In order to do that I had to include the header files

```cmake
    set(FETCHCONTENT_QUIET OFF)
    include(FetchContent)
    FetchContent_Declare(
            ortools
            GIT_REPOSITORY https://github.com/google/or-tools.git
            GIT_TAG        stable
    )
    set(BUILD_SHARED_LIBS ON)
    set(BUILD_DEPS ON)
    set(BUILD_SAMPLES OFF)
    set(BUILD_EXAMPLES OFF)
    FetchContent_MakeAvailable(ortools)
    add_definitions(-DORTOOLS)
    include_directories(${ortools_SOURCE_DIR})
    include_directories(${ortools_BINARY_DIR})
    include_directories(${absl_SOURCE_DIR})
    include_directories(${protobuf_SOURCE_DIR}/../src)
```

Then I'm building a bunch of static libraries, none of which are linked to ortools, but some of which include headers from ortools. 

Then I link all those static libraries together into a shared library and to ortools:
```
target_link_libraries(${shared_library} PRIVATE ${all_the_static_libraries} ortools::ortools)
```

Everything compiles, but when I go to run the program immediately crashes with the error:

ERROR: Something is wrong with flag 'flagfile' in file 'Users/connorriley/dial-a-ride-dispatcher/cmake-build-debug/_deps/absl-src/absl/flags/parse.cc'. One possibility: file 'Users/connorriley/dial-a-ride-dispatcher/cmake-build-debug/_deps/absl-src/absl/flags/parse.cc' is being linked both statically and dynamically into this executable. e.g. some files listed as srcs to a test and also listed as srcs of some shared lib deps of the same test.

The executable I'm running is linked as such:

```
add_executable(dispatcher main.cc)
target_link_libraries(dispatcher PRIVATE ${shared_library} ${Boost_LIBRARIES} OpenMP::OpenMP_CXX)
```

I'm really out of my skill set in terms of what I should be doing to try to debug this. Any suggestions you have would be appreciated.

**Comment by ctriley** [comment, marked as answer]
Update, I got my code compiling, I still don't really understand what was going wrong. The only thing I changed from above is instead of having 

```cmake
    set(BUILD_SHARED_LIBS ON)
```
I set it to 
```cmake
    set(BUILD_SHARED_LIBS OFF)
```

The thing that clued me in to this might being the issue was the discussion here: https://github.com/abseil/abseil-cpp/issues/125
which one of the maintainers of absl says:

"Unfortunately, Abseil has not been designed with use in dynamic libraries in mind (https://abseil.io/about/compatibility) and I don't know of a way to fix this issue in a way that's portable across all of the platforms that Abseil supports."

I'm not sure if that discussion is relevant, but if anyone else comes across this I figured I'd leave it here.

**Comment by rlorigro** [comment]
Hi, I was following along with your solution and noticed this:
```
    include_directories(${protobuf_SOURCE_DIR}/../src)
```

This will result in an incorrect path, because the `..` should not be involved. The `/src/` dir is a direct child of the `protobuf_SOURCE_DIR`. I was using the above answer and as a result my build was trying to find the protobuf header from my system location, causing a version mismatch, and compiler errors. I uninstalled my system level libprotobuf-dev because it was not needed by anything else, and removed the `..`, which fixed the issue. So the corrected solution looks like this:

```
include_directories(
        ${ortools_SOURCE_DIR}
        ${ortools_BINARY_DIR}
        ${absl_SOURCE_DIR}
        ${protobuf_SOURCE_DIR}/src
)
```

Just wanted to add this in case anyone else gets caught up by the same thing.
